riscv.md

```diff
@@ -416,6 +417,44 @@ typedef struct {
 } tls_index;
 ```
 
+## <a name=program-headers></a> Program Header Table
+
+The following table lists additional processor-specific segment types:
+
+Name                      | Value      | Description
+--------------------------|------------|-------------------------------------
+PT\_RISCV\_MEMTAG\_CHERI  | 0x7fffffff | Capability memory tags
+
+`PT_RISCV_MEMTAG_CHERI` segments (if present) hold capability memory tags
+for a particular memory range. They are defined for core dump files of
+type `ET_CORE`.  The program header for these segments is defined as:
+
+* `p_type`: `PT_RISCV_MEMTAG_CHERI`
+
+* `p_flags`: bitmask of capability tag MMU permissions
+
+* `p_offset`: segment file offset
+
+* `p_vaddr`: segment virtual address, must be page-aligned
+
+* `p_paddr`: 0
+
+* `p_filesz`: segment size in file, calculated as `p_memsz / (8 *
+  sizeof(void * __capability)` (eight 1-bit tags cover 8 capabilities of
+  memory)
+
+* `p_memsz`: segment size in memory, must be a multiple of the page size
+
+* `p_align`: 0
+
+The tags are stored in the core file at `p_offset` as eight 1-bit tags
+in a byte. With a capability size of 16 bytes, a 4K page requires 32
+bytes in the core file.
+
+``p_flags`` contains `PF_R` if the segment's MMU permissions permit reading
+capability tags and `PF_W` if the segment's MMU permissions permit writing
+capability tags.
```

Comment:
Is this here as a way to include LOAD_CAP/STORE_CAP?

Reply:
Yes.  I find it useful in readelf output, e.g.:

```
> readelf -l thread.core 

Elf file type is CORE (Core file)
Entry point 0x0
There are 78 program headers, starting at offset 64

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flg    Align
  NOTE           0x0000000000001150 0x0000000000000000 0x0000000000000000
                 0x00000000000072c4 0x0000000000000000  R      0x4
  LOAD           0x0000000000009000 0x0000000000121000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x000000000000a000 0x0000000000131000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x000000000000b000 0x0000000040171000 0x0000000000000000
                 0x0000000000003000 0x0000000000003000  RW     0x1000
  LOAD           0x000000000000e000 0x0000000040183000 0x0000000000000000
                 0x0000000000003000 0x0000000000003000  RW     0x1000
  LOAD           0x0000000000011000 0x0000000040186000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x0000000000012000 0x0000000040187000 0x0000000000000000
                 0x0000000000009000 0x0000000000009000  RW     0x1000
  LOAD           0x000000000001b000 0x0000000040190000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x000000000001c000 0x00000000401d3000 0x0000000000000000
                 0x0000000000002000 0x0000000000002000  R      0x1000
  LOAD           0x000000000001e000 0x00000000401d5000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x000000000001f000 0x00000000401e5000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x0000000000020000 0x00000000401e6000 0x0000000000000000
                 0x0000000000014000 0x0000000000014000  RW     0x1000
  LOAD           0x0000000000034000 0x00000000403e2000 0x0000000000000000
                 0x000000000001e000 0x000000000001e000  R      0x1000
  LOAD           0x0000000000052000 0x0000000040400000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x0000000000053000 0x0000000040410000 0x0000000000000000
                 0x000000000000b000 0x000000000000b000  RW     0x1000
  LOAD           0x000000000005e000 0x000000004041b000 0x0000000000000000
                 0x0000000000441000 0x0000000000441000  RW     0x1000
  LOAD           0x000000000049f000 0x000000004085c000 0x0000000000000000
                 0x0000000000004000 0x0000000000004000  RW     0x1000
  LOAD           0x00000000004a3000 0x0000000040868000 0x0000000000000000
                 0x0000000000028000 0x0000000000028000  RW     0x1000
  LOAD           0x00000000004cb000 0x0000000040890000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x00000000006cb000 0x0000000040a90000 0x0000000000000000
                 0x0000000000009000 0x0000000000009000  RW     0x1000
  LOAD           0x00000000006d4000 0x0000000040a99000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x00000000006d5000 0x0000000040a9a000 0x0000000000000000
                 0x0000000000010000 0x0000000000010000  RW     0x1000
  LOAD           0x00000000006e5000 0x0000000040e8a000 0x0000000000000000
                 0x0000000000020000 0x0000000000020000  RW     0x1000
  LOAD           0x0000000000705000 0x000000004128a000 0x0000000000000000
                 0x0000000000020000 0x0000000000020000  RW     0x1000
  LOAD           0x0000000000725000 0x00000000412aa000 0x0000000000000000
                 0x0000000000010000 0x0000000000010000  RW     0x1000
  LOAD           0x0000000000735000 0x000000004169a000 0x0000000000000000
                 0x0000000000020000 0x0000000000020000  RW     0x1000
  LOAD           0x0000000000755000 0x0000000041a9a000 0x0000000000000000
                 0x0000000000020000 0x0000000000020000  RW     0x1000
  LOAD           0x0000000000775000 0x0000000041aba000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000000975000 0x0000000041cba000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000000b75000 0x0000000041eba000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000000d75000 0x00000000420ba000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000000f75000 0x0000000060000000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000001175000 0x0000000080000000 0x0000000000000000
                 0x0000000000600000 0x0000000000600000  RW     0x1000
  LOAD           0x0000000001775000 0x00000000a0000000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000001975000 0x00000000c0000000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000001b75000 0x00000000e0000000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000001d75000 0x0000000100000000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000001f75000 0x0000ffffbff00000 0x0000000000000000
                 0x0000000000080000 0x0000000000080000  RW     0x1000
  LOAD           0x0000000001ff5000 0x0000fffffff60000 0x0000000000000000
                 0x0000000000020000 0x0000000000020000  RW     0x1000
  LOAD           0x0000000002015000 0x0000fffffffff000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  R E    0x1000
  AARCH64_MEMTAG 0x0000000002016000 0x0000000000121000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x0000000002016020 0x0000000000131000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x0000000002016040 0x0000000040171000 0x0000000000000000
                 0x0000000000000060 0x0000000000003000  RW     0
  AARCH64_MEMTAG 0x00000000020160a0 0x0000000040183000 0x0000000000000000
                 0x0000000000000060 0x0000000000003000  RW     0
  AARCH64_MEMTAG 0x0000000002016100 0x0000000040186000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x0000000002016120 0x0000000040187000 0x0000000000000000
                 0x0000000000000120 0x0000000000009000  RW     0
  AARCH64_MEMTAG 0x0000000002016240 0x0000000040190000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x0000000002016260 0x00000000401d3000 0x0000000000000000
                 0x0000000000000040 0x0000000000002000  R      0
  AARCH64_MEMTAG 0x00000000020162a0 0x00000000401d5000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x00000000020162c0 0x00000000401e5000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x00000000020162e0 0x00000000401e6000 0x0000000000000000
                 0x0000000000000280 0x0000000000014000  RW     0
  AARCH64_MEMTAG 0x0000000002016560 0x00000000403e2000 0x0000000000000000
                 0x00000000000003c0 0x000000000001e000  R      0
  AARCH64_MEMTAG 0x0000000002016920 0x0000000040400000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x0000000002016940 0x0000000040410000 0x0000000000000000
                 0x0000000000000160 0x000000000000b000  RW     0
  AARCH64_MEMTAG 0x0000000002016aa0 0x000000004041b000 0x0000000000000000
                 0x0000000000008820 0x0000000000441000  RW     0
  AARCH64_MEMTAG 0x000000000201f2c0 0x000000004085c000 0x0000000000000000
                 0x0000000000000080 0x0000000000004000  RW     0
  AARCH64_MEMTAG 0x000000000201f340 0x0000000040868000 0x0000000000000000
                 0x0000000000000500 0x0000000000028000  RW     0
  AARCH64_MEMTAG 0x000000000201f840 0x0000000040890000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002023840 0x0000000040a90000 0x0000000000000000
                 0x0000000000000120 0x0000000000009000  RW     0
  AARCH64_MEMTAG 0x0000000002023960 0x0000000040a99000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x0000000002023980 0x0000000040a9a000 0x0000000000000000
                 0x0000000000000200 0x0000000000010000  RW     0
  AARCH64_MEMTAG 0x0000000002023b80 0x0000000040e8a000 0x0000000000000000
                 0x0000000000000400 0x0000000000020000  RW     0
  AARCH64_MEMTAG 0x0000000002023f80 0x000000004128a000 0x0000000000000000
                 0x0000000000000400 0x0000000000020000  RW     0
  AARCH64_MEMTAG 0x0000000002024380 0x00000000412aa000 0x0000000000000000
                 0x0000000000000200 0x0000000000010000  RW     0
  AARCH64_MEMTAG 0x0000000002024580 0x000000004169a000 0x0000000000000000
                 0x0000000000000400 0x0000000000020000  RW     0
  AARCH64_MEMTAG 0x0000000002024980 0x0000000041a9a000 0x0000000000000000
                 0x0000000000000400 0x0000000000020000  RW     0
  AARCH64_MEMTAG 0x0000000002024d80 0x0000000041aba000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002028d80 0x0000000041cba000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x000000000202cd80 0x0000000041eba000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002030d80 0x00000000420ba000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002034d80 0x0000000060000000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002038d80 0x0000000080000000 0x0000000000000000
                 0x000000000000c000 0x0000000000600000  RW     0
  AARCH64_MEMTAG 0x0000000002044d80 0x00000000a0000000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002048d80 0x00000000c0000000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x000000000204cd80 0x00000000e0000000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002050d80 0x0000000100000000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002054d80 0x0000ffffbff00000 0x0000000000000000
                 0x0000000000001000 0x0000000000080000  RW     0
  AARCH64_MEMTAG 0x0000000002055d80 0x0000fffffff60000 0x0000000000000000
                 0x0000000000000400 0x0000000000020000  RW     0
 ```

Here the 'RW' tags on `AARCH64_MEMTAG` correspond to `VM_PROT_READ_CAP` and `VM_PROT_WRITE_CAP`.

(It probably goes without saying, but PT_RISCV_MEMTAG_CHERI and PT_AARCH64_MEMTAG_CHERI are identical aside from the arch-specific PT_* value).